llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2787,9 +2787,16 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
                                     GEP.getNoWrapFlags()));
   }
 
-  // Canonicalize scalable GEPs to an explicit offset using the llvm.vscale
-  // intrinsic. This has better support in BasicAA.
-  if (GEPEltType->isScalableTy()) {
+  // Canonicalize
+  //  - scalable GEPs to an explicit offset using the llvm.vscale intrinsic.
+  //    This has better support in BasicAA.
+  //  - gep i32 p, mul(O, C) -> gep i8, p, mul(O, C*4) to fold the two
+  //    multiplies together.
+  if (GEPEltType->isScalableTy() ||
+      (!GEPEltType->isIntegerTy(8) && GEP.getNumIndices() == 1 &&
+       match(GEP.getOperand(1),
+             m_OneUse(m_CombineOr(m_Mul(m_Value(), m_ConstantInt()),
+                                  m_Shl(m_Value(), m_ConstantInt())))))) {
```

Comment:
Not sure if it comes up but you could also have `exact` `div`/`shr`

Reply:
Because it can fold into the mul? Thanks - I can take a look.

But I would prefer to do it in a separate patch if that is OK. If I rebase I can check if it comes up in practice. (Or I could just implement it anyway, let me know if you think that is better).

Reply:
We already handle the exact div/shr case separately.

Reply:
We dont seem to: https://godbolt.org/z/1xhh1h5e9 (but we could: https://alive2.llvm.org/ce/z/QvqkHE)

Reply:
We handle the case where the div and mul cancel. I guess it makes sense to extend it to the case where the div constant is a multiple of the mul constant.